overlays/ghc-packages.nix

```diff
@@ -65,38 +149,77 @@ let
+  ghc-boot-packages = builtins.mapAttrs
+    (ghcName: value: builtins.mapAttrs
+      (pkgName: sdistAndNix: importSdistAndNix {
+        inherit (sdistAndNix) sdist;
+        nix = ghc-boot-packages-nix."${ghcName}" + "/${pkgName}.nix";
+      }) value)
+        ghc-boot-packages-sdist-and-nix;
+
+  # Derivation with cabal.project for use with `cabalProject'` for each ghc. 
   ghc-extra-pkgs-cabal-projects = builtins.mapAttrs (name: value:
     let package-locs =
-      builtins.mapAttrs (_: dir: "${value.passthru.configured-src}/${dir}")
         # TODO ghc-heap.cabal requires cabal 3.  We should update the cabalProject' call
         # in `ghc-extra-projects` below to work with this.
         (self.lib.filterAttrs (n: _: n != "base" && n != "ghc-heap") ghc-extra-pkgs);
-    in self.writeTextFile {
+    in self.stdenv.mkDerivation {
       name = "ghc-extra-pkgs-cabal-project-${name}";
-      destination = "/cabal.project";
-      text = ''
+      phases = [ "buildPhase" ];
+      # Copy each cabal file from the configured ghc source and
+      # add a suitable cabal.project file.
+      buildPhase = ''
+        ${self.lib.concatStrings (self.lib.mapAttrsToList (_: dir: ''
+          mkdir -p $out/${dir}
+          cp ${value.passthru.configured-src}/${dir}/*.cabal $out/${dir}
+          # Remove references to libffi as the are not cross platform
+          # and will break memoization (we will need to add them back)
+          sed -i 's|/nix/store/.*-libffi.*/include||' $out/${dir}/*.cabal
+        '') package-locs)}
+        cat >$out/cabal.project <<EOF
         packages: ${self.lib.concatStringsSep " " (self.lib.attrValues package-locs)}
         -- need this for libiserve as it doesn't build against 3.0 yet.
         constraints: network < 3.0,
                      ghc +ghci,
                      ghci +ghci,
                      libiserv +network
+        EOF
       '';
     }) self.buildPackages.haskell-nix.compiler;
 
-  ghc-extra-projects = builtins.mapAttrs (name: proj: self.haskell-nix.cabalProject' {
-      name = "ghc-extra-packages";
+  # A `cabalProject'` project for each ghc
+  ghc-extra-projects = builtins.mapAttrs (ghcName: proj:
+    # Where to look for materialization files
+    let materializedPath = ../materialized/ghc-extra-projects
+                             + "/${ghc-extra-projects-type}/${ghcName}";
+    in self.haskell-nix.cabalProject' {
+      name = "ghc-extra-projects-${ghc-extra-projects-type}-${ghcName}";
       src = proj;
-      index-state = "2019-10-31T00:00:00Z";
-      ghc = self.buildPackages.haskell-nix.compiler.${name};
+      index-state = "2020-04-12T00:00:00Z";
+      plan-sha256 =
+        ghc-extra-projects-sha256."${ghc-extra-projects-type}"."${ghcName}"
+          or null;
+      materialized =
+        if ghc-extra-projects-sha256."${ghc-extra-projects-type}" ? "${ghcName}"
+            && __pathExists materializedPath
```

Comment:
Maybe we should move this check into `materialize` so we don't have to do it at the call sites?